18F/identity-idp
# LG-10479 FSM/SSN - Check redirect for SSN Step

## 🎫 Ticket
[LG-10479](https://cm-jira.usa.gov/browse/LG-10479) FSM/SSN - Check redirect for SSN Step

## 🛠 Summary of changes
* When  the `in_person_ssn_info_controller_enabled` flag is set to true and when flow_session.pii_from_user has a ssn, mark the in person FSM ssn step as complete. (The user visited the non-FSM SSN step but we still need to mark the FSM SSN step as complete while we still have the FSM ssn step.)

## 📜 Testing Plan
- [ ] Check that the bug  🐜 found in [PR#8844](https://github.com/18F/identity-idp/pull/8844) is resolved:
(You can test on Joy or you can watch the before and after demo videos below)
1. The `in_person_ssn_info_controller_enabled` flag should be set to true in the environment you will be testing.
2. Create an account
3. Go to /verify
4. Upload state-issued ID (front + back) to fail ID verification
5. Go to in person proofing
6. Fill in State-Id information, Residential Address, and Social Security Number. (Notice the url for SSN is `/verify/in_person_proofing/ssn`. This is the non-FSM SSN version.)
7. Once on `/verify/in_person/verify_info` change the url to the FSM version of SSN (`/verify/in_person/ssn`). You should not be redirected back to the SSN. You should remain on the Verify Info page.  (This is what we fixed. Previously you would be directed back).
8. Confirm you can visit the SSN page again my clicking Update on the SSN.  You should be able to go back. (The url should be the new non-FSM url `/verify/in_person_proofing/ssn`.
9. Once you are back on Verify Info (`/verify/in_person/verify_info`) change the url to the non-FSM version of SSN (`/verify/in_person_proofing/ssn`). You should not be redirected back to the SSN. You should remain on the Verify Info page.  (This is to confirm the same behavior occurs for the non-FSM step).
10. Confirm you can visit the SSN page again my clicking Update on the SSN.  You should be able to go back. (The url should be the new non-FSM url `/verify/in_person_proofing/ssn`.
- [ ] Check that user cannot hit ssn step by typing in url before they have arrived at this step by walking through the flow for both FSM and non-FSM flows 
- [ ] After arriving at the ssn step [by walking through the flow] the user cannot return to the ssn step by typing in the url for both FSM and non-FSM flows

### Important Notes 📓 
- non-FSM can be tested in [joy](https://idp.joy.identitysandbox.gov/account)- it has the `in_person_ssn_info_controller_enabled` flag is set to true 

- fsm can be tested in [dev](https://idp.dev.identitysandbox.gov/)- it has the `in_person_ssn_info_controller_enabled` flag is set to false

## 👀 Demo

<details>
<summary>Before: -- Please Open --</summary>
You should not be redirected so the SSN view when you change the URL to `/verify/in_person/ssn` on the Verify your information page but you are. This is what we are solving for. 

- [ ] - @0:11 seconds - Url is `/verify/in_person_proofing/ssn` - This shows the feature flag is on and we are using the non-FSM SSN step.
- [ ] - @0:19 seconds - Url is changed to `/verify/in_person/ssn` (FSM-step) and we are redirected- we should not be!
- [ ] - @0:42 seconds - Url is changed to `/verify/in_person_proofing/ssn` (non-FSM step) and we are not redirected (as expected).

https://github.com/18F/identity-idp/assets/125507397/f6579081-aa0f-4dec-bab9-46acd8a3dd6b

</details>

<details>
<summary>After:  -- Please Open --</summary>
You should not be redirected so the SSN view when you change the URL to `/verify/in_person/ssn` on the Verify your information page, problem fixed. 

- [ ] - @0:11 seconds - Url is `/verify/in_person_proofing/ssn` - This shows the feature flag is on and we are using the non-FSM SSN step.
- [ ] - @0:21 seconds - Url is changed to `/verify/in_person/ssn` (FSM-step) and we are not redirected- this is what was fixed!
- [ ] - @0:45 seconds - Url is changed to `/verify/in_person_proofing/ssn` (non-FSM step) and we are not redirected (as expected).

https://github.com/18F/identity-idp/assets/125507397/c8ef5da6-782e-4ec1-a2c9-4774de8d269b

</details>
